### Spark HW2 Moive Recommendation
In this notebook, we will use an Alternating Least Squares (ALS) algorithm with Spark APIs to predict the ratings for the movies in [MovieLens small dataset](https://grouplens.org/datasets/movielens/latest/)

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import math
%matplotlib inline

%matplotlib inline is not supported in Databricks.
You can display matplotlib figures using display(). For an example, see https://docs.databricks.com/user-guide/visualizations/matplotlib-and-ggplot.html

In [3]:
import os
os.environ["PYSPARK_PYTHON"] = "python3"

## Part1: Data ETL and Data Exploration

In [5]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("moive analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [6]:
movies_df = spark.read.load("/FileStore/tables/movies.csv", format='csv', header = True)
ratings_df = spark.read.load("/FileStore/tables/ratings.csv", format='csv', header = True)
links_df = spark.read.load("/FileStore/tables/links.csv", format='csv', header = True)
tags_df = spark.read.load("/FileStore/tables/tags.csv", format='csv', header = True)

In [7]:
movies_df.show(5)

+-------+--------------------+--------------------+
movieId| title| genres|
+-------+--------------------+--------------------+
 1| Toy Story (1995)|Adventure|Animati...|
 2| Jumanji (1995)|Adventure|Childre...|
 3|Grumpier Old Men ...| Comedy|Romance|
 4|Waiting to Exhale...|Comedy|Drama|Romance|
 5|Father of the Bri...| Comedy|
+-------+--------------------+--------------------+
only showing top 5 rows

In [8]:
ratings_df.show(5)

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
+------+-------+------+---------+
only showing top 5 rows

In [9]:
links_df.show(5)

+-------+-------+------+
movieId| imdbId|tmdbId|
+-------+-------+------+
 1|0114709| 862|
 2|0113497| 8844|
 3|0113228| 15602|
 4|0114885| 31357|
 5|0113041| 11862|
+-------+-------+------+
only showing top 5 rows

In [10]:
tags_df.show(5)

+------+-------+---------------+----------+
userId|movieId| tag| timestamp|
+------+-------+---------------+----------+
 2| 60756| funny|1445714994|
 2| 60756|Highly quotable|1445714996|
 2| 60756| will ferrell|1445714992|
 2| 89774| Boxing story|1445715207|
 2| 89774| MMA|1445715200|
+------+-------+---------------+----------+
only showing top 5 rows

In [11]:
tmp1 = ratings_df.groupBy("userID").count().toPandas()['count'].min()
tmp2 = ratings_df.groupBy("movieId").count().toPandas()['count'].min()
print('For the users that rated movies and the movies that were rated:')
print('Minimum number of ratings per user is {}'.format(tmp1))
print('Minimum number of ratings per movie is {}'.format(tmp2))

For the users that rated movies and the movies that were rated:
Minimum number of ratings per user is 20
Minimum number of ratings per movie is 1

In [12]:
tmp1 = sum(ratings_df.groupBy("movieId").count().toPandas()['count'] == 1)
tmp2 = ratings_df.select('movieId').distinct().count()
print('{} out of {} movies are rated by only one user'.format(tmp1, tmp2))

3446 out of 9724 movies are rated by only one user

## Part 1: Spark SQL and OLAP

In [14]:
movies_df.registerTempTable("movies")
ratings_df.registerTempTable("ratings")
links_df.registerTempTable("links")
tags_df.registerTempTable("tags")

### Q1: The number of Users

In [16]:
%sql select count(distinct userID) from ratings

count(DISTINCT userID) 610

### Q2: The number of Movies

In [18]:
%sql select count(distinct movieID) from movies

count(DISTINCT movieID) 9742

### Q3:  How many movies are rated by users? List movies not rated before

In [20]:
%sql select title, genres from movies where movieID not in (select distinct movieID from ratings)

title genres Innocents, The (1961) Drama|Horror|Thriller Niagara (1953) Drama|Thriller For All Mankind (1989) Documentary Color of Paradise, The (Rang-e khoda) (1999) Drama I Know Where I'm Going! (1945) Drama|Romance|War Chosen, The (1981) Drama Road Home, The (Wo de fu qin mu qin) (1999) Drama|Romance Scrooge (1970) Drama|Fantasy|Musical Proof (1991) Comedy|Drama|Romance Parallax View, The (1974) Thriller This Gun for Hire (1942) Crime|Film-Noir|Thriller Roaring Twenties, The (1939) Crime|Drama|Thriller Mutiny on the Bounty (1962) Adventure|Drama|Romance In the Realms of the Unreal (2004) Animation|Documentary Twentieth Century (1934) Comedy Call Northside 777 (1948) Crime|Drama|Film-Noir Browning Version, The (1951) Drama Chalet Girl (2011) Comedy|Romance

In [21]:
%sql select count(distinct movieID) from ratings 

count(DISTINCT movieID) 9724

### Q4: List Movie Genres

In [23]:
%sql select distinct genres from movies

genres Comedy|Horror|Thriller Adventure|Sci-Fi|Thriller Action|Adventure|Drama|Fantasy Action|Drama|Horror Action|Animation|Comedy|Sci-Fi Animation|Children|Drama|Musical|Romance Action|Adventure|Drama Adventure|Sci-Fi Documentary|Musical|IMAX Adventure|Children|Fantasy|Sci-Fi|Thriller Adventure|Animation Musical|Romance|War Action|Adventure|Fantasy|Romance Adventure|Children|Drama|Fantasy|IMAX Comedy|Crime|Horror|Thriller Crime|Drama|Fantasy|Horror|Thriller Comedy|Mystery|Thriller Adventure|Fantasy Horror|Romance|Sci-Fi Drama|Film-Noir|Romance Crime|Drama|Film-Noir|Mystery|Thriller Musical|Western Adventure|Animation|Children|Musical|Romance Comedy|Crime|Mystery|Romance|Thriller Action|Animation|Children|Crime Adventure|Horror|Sci-Fi Comedy|Drama|Romance|Thriller Action|Drama|Mystery|Sci-Fi|Thriller Adventure|Animation|Children|Comedy|Drama|Fantasy Action|Adventure|Mystery|Sci-Fi Adventure|Drama|Romance|Thriller|War Animation|Drama Horror|Thriller|IMAX Crime Action|Animation|Film-Noir|Sci-Fi|Thriller Action|Crime|Horror|Sci-Fi|Thriller Crime|Drama|Thriller Action|Adventure|Animation|Sci-Fi Action|Adventure|Drama|Fantasy|Mystery|IMAX Mystery|Sci-Fi|Thriller Adventure|Drama|IMAX Romance Comedy|Horror|Musical|Sci-Fi Children|Drama|Musical Action|Drama Crime|Thriller Drama|Mystery|Sci-Fi|Thriller Comedy|Crime|Drama|Western Action|Adventure|Drama|Fantasy|Romance|Sci-Fi|Thriller Action|Fantasy|Horror|IMAX Action|Sci-Fi Adventure|Crime|Drama|Romance Drama|Fantasy|Mystery|Sci-Fi Action|Crime|Mystery Animation|Children|Comedy|Fantasy|Musical|Romance|IMAX Action|Adventure|Sci-Fi|War|IMAX Comedy|Drama|Fantasy|Mystery|Romance Children|Comedy|Crime|Drama|Fantasy Action|Comedy|Fantasy|IMAX Action|Adventure|Fantasy Thriller Comedy|Fantasy|Romance Romance|Thriller Adventure|Children|Comedy|Mystery Drama|Fantasy|Romance|Sci-Fi Documentary|Drama|Thriller Crime|Drama|Horror|Mystery|Thriller Action|Romance|Thriller Action|Crime|Drama|Sci-Fi|Thriller Film-Noir|Mystery|Thriller Crime|Thriller|War Comedy|Drama|Fantasy|Romance Adventure|Animation|Children|Comedy|Fantasy|Romance Comedy|Crime|Drama|War Adventure|Animation|Fantasy|Romance Drama|Fantasy|Horror|Romance|Thriller Action|Sci-Fi|Thriller|Western|IMAX Animation|Children Action|Crime|Mystery|Thriller Comedy|Romance|Sci-Fi Adventure|War|Western Action|Animation|Comedy|Fantasy Comedy|Horror|Romance Comedy|Drama|Musical Drama|Film-Noir|Mystery|Romance Drama|Musical|Romance Adventure|Comedy|Thriller Action|Animation|Children|Comedy|Sci-Fi|IMAX Drama|Sci-Fi|IMAX Animation|Children|Mystery Action|Sci-Fi|Thriller|Western Action|Adventure|Animation|Comedy|Crime|Mystery Action|Adventure|Animation|Comedy|Fantasy|Mystery|Sci-Fi Adventure|Comedy|Drama|Romance Action|Adventure|Animation|Children|Comedy|Western Comedy|Horror Animation|Children|Comedy|Romance Animation|Children|Comedy|Musical|Romance Adventure|Animation|Fantasy Children|Comedy|Crime|Musical Action|Adventure|Drama|Sci-Fi|Thriller Action|Comedy|Crime|Mystery Crime|Horror|Thriller Adventure|Comedy|Sci-Fi Animation|Children|Comedy|Fantasy|Musical|Romance Action|Adventure|Drama|Sci-Fi Action|Drama|Horror|Thriller Adventure|Children Animation|Fantasy|Thriller Comedy|Drama|Horror|Sci-Fi|Thriller Action|Thriller|War Action|Crime|Mystery|Sci-Fi|Thriller Adventure|Animation|Children|Musical|Western Crime|Drama|Fantasy|Mystery|Thriller Animation|Comedy|Horror|IMAX Action|Comedy|Crime|Fantasy Adventure|Animation|Comedy|Crime Action|Adventure|Fantasy|Thriller Action|Adventure|Fantasy|Horror Action|Animation|Sci-Fi Action|Adventure|Drama|Romance|Thriller|Western Adventure|Comedy|Fantasy|Musical Action|Adventure|Animation|Children|Comedy|Sci-Fi Action|Adventure|Crime|IMAX Action|Crime|Thriller Crime|Drama Adventure|Comedy|Western Documentary|Drama Romance|Sci-Fi Adventure|Drama|Horror|Sci-Fi|Thriller Drama|Sci-Fi|War Action|Comedy|Drama|Horror|Thriller Action|Comedy|Crime|Horror Comedy|Mystery|Romance Comedy|Musical|Romance Fantasy|Sci-Fi Action|Animation|

In [24]:
%sql select distinct explode(split(genres, '[|]')) as Genres from movies order by Genres 

Genres (no genres listed) Action Adventure Animation Children Comedy Crime Documentary Drama Fantasy Film-Noir Horror IMAX Musical Mystery Romance Sci-Fi Thriller War Western

### Q5: Movie for Each Category

In [26]:
%sql select title, explode(split(genres, '[|]')) as Genres from movies order by Genres 

title Genres Too Funny to Fail: The Life and Death of The Dana Carvey Show (2017) (no genres listed) Superfast! (2015) (no genres listed) Ali Wong: Baby Cobra (2016) (no genres listed) Serving in Silence: The Margarethe Cammermeyer Story (1995) (no genres listed) A Christmas Story Live! (2017) (no genres listed) Green Room (2015) (no genres listed) A Midsummer Night's Dream (2016) (no genres listed) The Forbidden Dance (1990) (no genres listed) Ben-hur (2016) (no genres listed) The Brand New Testament (2015) (no genres listed) Ethel & Ernest (2016) (no genres listed) La cravate (1957) (no genres listed) Whiplash (2013) (no genres listed) Let It Be Me (1995) (no genres listed) Hyena Road (no genres listed) The OA (no genres listed) Lemonade (2016) (no genres listed) The Adventures of Sherlock Holmes and Doctor Watson (no genres listed) Cosmos (no genres listed) Maria Bamford: Old Baby (no genres listed) Trevor Noah: African American (2013) (no genres listed) A Cosmic Christmas (1977) (no genres listed) Death Note: Desu nôto (2006–2007) (no genres listed) Pirates of the Caribbean: Dead Men Tell No Tales (2017) (no genres listed) Generation Iron 2 (no genres listed) Grease Live (2016) (no genres listed) T2 3-D: Battle Across Time (1996) (no genres listed) The Godfather Trilogy: 1972-1990 (1992) (no genres listed) Guardians (2016) (no genres listed) Noin 7 veljestä (1968) (no genres listed) The Adventures of Sherlock Holmes and Doctor Watson: The Hunt for the Tiger (1980) (no genres listed) The Putin Interviews (2017) (no genres listed) Paterson (no genres listed) Black Mirror (no genres listed) Troy (2004) Action Bad Boys (1995) Action Net, The (1995) Action Enter the Dragon (1973) Action Congo (1995) Action Wages of Fear, The (Salaire de la peur, Le) (1953) Action New York Cop (Nyû Yôku no koppu) (1993) Action Tales from the Hood (1995) Action Rob Roy (1995) Action You Only Live Twice (1967) Action Hackers (1995) Action Under Siege 2: Dark Territory (1995) Action Drop Zone (1994) Action Jason and the Argonauts (1963) Action Quick and the Dead, The (1995) Action Legend, The (Legend of Fong Sai-Yuk, The) (Fong Sai Yuk) (1993) Action I Love Trouble (1994) Action Speed (1994) Action Dobermann (1997) Action Beverly Hills Cop III (1994) Action Die Hard: With a Vengeance (1995) Action Never Say Never Again (1983) Action Judge Dredd (1995) Action Program, The (1993) Action Delta Force, The (1986) Action Four Musketeers, The (1974) Action Hunted, The (1995) Action Heavy Metal (1981) Action Outbreak (1995) Action Sniper (1993) Action Stargate (1994) Action Eraser (1996) Action Clear and Present Danger (1994) Action Master of the Flying Guillotine (Du bi quan wang da po xue di zi) (1975) Action Naked Gun 33 1/3: The Final Insult (1994) Action Bridge Too Far, A (1977) Action True Lies (1994) Action Low Down Dirty Shame, A (1994) Action Street Fighter (1994) Action Shiri (Swiri) (1999) Action Cliffhanger (1993) Action Demolition Man (1993) Action Octopussy (1983) Action In the Line of Fire (1993) Action Jurassic Park (1993) Action Menace II Society (1993) Action Versus (2000) Action Lupin III: The Castle Of Cagliostro (Rupan sansei: Kariosutoro no shiro) (1979) Action RoboCop 3 (1993) Action Shadow, The (1994) Action Striking Distance (1993) Action Surviving the Game (1994) Action Three Musketeers, The (1993) Action Pagemaster, The (1994) Action Alexander Nevsky (Aleksandr Nevskiy) (1938) Action Northwest Passage (1940) Action Courage Under Fire (1996) Action Dragonheart (1996) Action Throne of Blood (Kumonosu jô) (1957) Action Substitute, The (1996) Action Rock, The (1996) Action Barb Wire (1996) Action Warriors, The (1979) Action Taxi 3 (2003) Action Fled (1996) Action Crow: City of Angels, The (1996) Action Duck, You Sucker (1971) Action Supercop 2 (Project S) (Chao ji ji hua) (1993) Action North by Northwest (1959) Action Maximum Risk (1996) Action Hidden Fortress, The (Kakushi-toride no san-akunin) (1958) Action Die Hard (1988) Action L

In [27]:
%sql select a.Genres, count(*) as Count from (select explode(split(genres, '[|]')) as Genres from movies) as a group by a.Genres order by a.Genres

Genres Count (no genres listed) 34 Action 1828 Adventure 1263 Animation 611 Children 664 Comedy 3756 Crime 1199 Documentary 440 Drama 4361 Fantasy 779 Film-Noir 87 Horror 978 IMAX 158 Musical 334 Mystery 573 Romance 1596 Sci-Fi 980 Thriller 1894 War 382 Western 167

## Part2: Spark ALS based approach for training model
We will use an Spark ML to predict the ratings, so let's reload "ratings.csv" using ``sc.textFile`` and then convert it to the form of (user, item, rating) tuples.

In [29]:
ratings_df.show()

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
 1| 70| 3.0|964982400|
 1| 101| 5.0|964980868|
 1| 110| 4.0|964982176|
 1| 151| 5.0|964984041|
 1| 157| 5.0|964984100|
 1| 163| 5.0|964983650|
 1| 216| 5.0|964981208|
 1| 223| 3.0|964980985|
 1| 231| 5.0|964981179|
 1| 235| 4.0|964980908|
 1| 260| 5.0|964981680|
 1| 296| 3.0|964982967|
 1| 316| 3.0|964982310|
 1| 333| 5.0|964981179|
 1| 349| 4.0|964982563|
+------+-------+------+---------+
only showing top 20 rows

In [30]:
movie_ratings0=ratings_df.drop('timestamp')

In [31]:
# Data type convert
from pyspark.sql.types import IntegerType, FloatType
movie_ratings0 = movie_ratings0.withColumn("userId", movie_ratings0["userId"].cast(IntegerType()))
movie_ratings0 = movie_ratings0.withColumn("movieId", movie_ratings0["movieId"].cast(IntegerType()))
movie_ratings0 = movie_ratings0.withColumn("rating", movie_ratings0["rating"].cast(FloatType()))

In [32]:
movie_ratings0.show()

+------+-------+------+
userId|movieId|rating|
+------+-------+------+
 1| 1| 4.0|
 1| 3| 4.0|
 1| 6| 4.0|
 1| 47| 5.0|
 1| 50| 5.0|
 1| 70| 3.0|
 1| 101| 5.0|
 1| 110| 4.0|
 1| 151| 5.0|
 1| 157| 5.0|
 1| 163| 5.0|
 1| 216| 5.0|
 1| 223| 3.0|
 1| 231| 5.0|
 1| 235| 4.0|
 1| 260| 5.0|
 1| 296| 3.0|
 1| 316| 3.0|
 1| 333| 5.0|
 1| 349| 4.0|
+------+-------+------+
only showing top 20 rows

In [33]:
from pyspark.sql.functions import monotonically_increasing_id
movie_ratings0 = movie_ratings0.withColumn("id", monotonically_increasing_id())
movie_ratings0.show()

+------+-------+------+---+
userId|movieId|rating| id|
+------+-------+------+---+
 1| 1| 4.0| 0|
 1| 3| 4.0| 1|
 1| 6| 4.0| 2|
 1| 47| 5.0| 3|
 1| 50| 5.0| 4|
 1| 70| 3.0| 5|
 1| 101| 5.0| 6|
 1| 110| 4.0| 7|
 1| 151| 5.0| 8|
 1| 157| 5.0| 9|
 1| 163| 5.0| 10|
 1| 216| 5.0| 11|
 1| 223| 3.0| 12|
 1| 231| 5.0| 13|
 1| 235| 4.0| 14|
 1| 260| 5.0| 15|
 1| 296| 3.0| 16|
 1| 316| 3.0| 17|
 1| 333| 5.0| 18|
 1| 349| 4.0| 19|
+------+-------+------+---+
only showing top 20 rows

In [34]:
movie_ratings0.count()

Out[17]: 100836

In [35]:
from pyspark.sql.functions import col
(movie_ratings, rest) = movie_ratings0.randomSplit([0.1, 0.9])
movie_ratings.count()

Out[18]: 10057

### ALS Model Selection and Evaluation

With the ALS model, we can use a grid search to find the optimal hyperparameters.

In [37]:
# import package
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.ml.tuning import CrossValidator,ParamGridBuilder

In [38]:
#Create test and train set
(training,test)=movie_ratings.randomSplit([0.8,0.2])
training.show()

+------+-------+------+---+
userId|movieId|rating| id|
+------+-------+------+---+
 1| 50| 5.0| 4|
 1| 673| 3.0| 39|
 1| 1298| 5.0| 90|
 1| 1408| 3.0| 94|
 1| 2596| 5.0|168|
 1| 2616| 4.0|169|
 1| 2797| 4.0|182|
 1| 3034| 5.0|199|
 2| 333| 4.0|233|
 2| 86345| 4.0|249|
 2| 91658| 2.5|252|
 3| 1587| 4.5|275|
 3| 70946| 5.0|298|
 4| 351| 3.0|325|
 4| 910| 5.0|353|
 4| 919| 5.0|356|
 4| 1060| 2.0|363|
 4| 1197| 5.0|377|
 4| 1199| 2.0|379|
 4| 1211| 1.0|381|
+------+-------+------+---+
only showing top 20 rows

In [39]:
training.count()

Out[21]: 8080

In [40]:
#Create ALS model 
als = ALS(regParam=0.01, userCol="userId", itemCol="movieId", ratingCol="rating", coldStartStrategy="drop")

In [41]:
type(als)

Out[23]: pyspark.ml.recommendation.ALS

In [42]:
#Tune model using ParamGridBuilder
param_grid = ParamGridBuilder() \
            .addGrid(als.rank, [5, 10]) \
            .addGrid(als.maxIter, [5, 10]) \
            .addGrid(als.regParam, [0.01, 0.1]) \
            .build()
print(len(param_grid))

8

In [43]:
# Define evaluator as RMSE
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")

In [44]:
# Build Cross validation b
cv = CrossValidator(estimator=als, estimatorParamMaps=param_grid, evaluator=evaluator, numFolds = 5)

In [45]:
#Fit ALS model to training data
cvModel=cv.fit(training)

/databricks/spark/python/pyspark/ml/util.py:791: UserWarning: Can not find mlflow. To enable mlflow logging, install MLflow library from PyPi.
 warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)

In [46]:
#Extract best model from the tuning exercise using ParamGridBuilder
best_model = cvModel.bestModel

### Model testing
And finally, make a prediction and check the testing error.

In [48]:
#Generate predictions and evaluate using RMSE
predictions=best_model.transform(test)
rmse = evaluator.evaluate(predictions)

In [49]:
#Print evaluation metrics and model parameters
print ("RMSE = "+str(rmse))
print ("**Best Model**")
print (" Rank: " + str(best_model.rank)), 
print (" MaxIter: " + str(best_model
    ._java_obj     # Get Java object
    .parent()      # Get parent (ALS estimator)
    .getMaxIter())), 
print (" RegParam: " + str(best_model._java_obj.parent().getRegParam())), 

RMSE = 1.3501344809650004
**Best Model**
 Rank: 10
 MaxIter: 10
 RegParam: 0.1
Out[30]: (None,)

In [50]:
predictions.show()

+------+-------+------+-----+----------+
userId|movieId|rating| id|prediction|
+------+-------+------+-----+----------+
 542| 1580| 3.5|84436| 3.9162242|
 560| 1580| 4.0|86184| 3.2760768|
 354| 1580| 3.5|53654| 3.1186962|
 115| 1645| 4.0|18200| 3.3766272|
 28| 3175| 1.5| 4394| 2.1234705|
 590| 3175| 2.5|90913| 2.2936256|
 135| 3175| 4.0|20557| 3.3230414|
 132| 8638| 4.0|20214| 1.776117|
 298| 36525| 2.5|45034| 1.6030469|
 414| 540| 2.0|62534| 1.72446|
 580| 858| 4.5|89377| 4.577468|
 222| 858| 4.5|32622| 3.9388347|
 122| 858| 5.0|18861| 4.8584895|
 279| 858| 4.0|41084| 4.3723955|
 590| 858| 5.0|90609| 3.3115263|
 135| 858| 4.0|20392| 3.9662604|
 387| 1127| 4.5|59389| 2.0397623|
 591| 1127| 3.0|91251| 1.8349891|
 222| 1721| 2.5|32656| 2.8459404|
 280| 1721| 4.0|41286| 2.1165035|
+------+-------+------+-----+----------+
only showing top 20 rows

### Model apply and see the performance

In [52]:
alldata = best_model.transform(movie_ratings)
rmse = evaluator.evaluate(alldata)
print ("RMSE = "+str(rmse))
alldata.count()

RMSE = 0.5644455121358646
Out[32]: 9613

In [53]:
alldata.registerTempTable("alldata")

In [54]:
%sql select * from alldata order by id

userId movieId rating id prediction 1 50 5.0 4 4.8139777 1 673 3.0 39 3.0140967 1 1298 5.0 90 4.787555 1 1408 3.0 94 3.1122477 1 2596 5.0 168 4.868236 1 2616 4.0 169 3.8497894 1 2628 4.0 171 1.4667412 1 2797 4.0 182 4.071091 1 3034 5.0 199 4.866223 1 3053 5.0 201 2.9049928 1 3439 4.0 211 1.7064148 2 318 3.0 232 3.6837125 2 333 4.0 233 3.9403274 2 86345 4.0 249 3.8730474 2 91658 2.5 252 2.4990103 3 1587 4.5 275 4.405271 3 70946 5.0 298 4.959308 4 319 5.0 321 -0.6944257 4 351 3.0 325 3.009282 4 902 4.0 350 0.32799235 4 910 5.0 353 4.625452 4 919 5.0 356 4.657628 4 1060 2.0 363 2.0111842 4 1197 5.0 377 4.4705815 4 1199 2.0 379 2.1057947 4 1211 1.0 381 1.1924 4 1213 4.0 382 3.7148514 4 1732 4.0 406 3.8900495 4 1966 2.0 420 1.9717973 4 2351 4.0 436 3.9435947 4 2571 1.0 442 1.5312067 4 2858 5.0 454 2.8146186 4 3175 1.0 467 1.1713831 4 3358 3.0 472 3.0967898 5 364 3.0 540 2.9546652 5 592 3.0 554 2.9313312 6 13 3.0 569 3.0122824 6 16 4.0 571 4.0046167 6 60 4.0 591 3.804031 6 66 3.0 595 2.9374733 6 76 4.0 596 3.517944 6 100 3.0 605 2.966446 6 113 3.0 611 2.949655 6 145 4.0 616 3.9295845 6 153 3.0 620 2.9510832 6 177 3.0 631 2.949655 6 185 3.0 635 3.0158434 6 209 3.0 646 2.949655 6 230 5.0 656 4.768962 6 234 2.0 658 2.086646 6 236 4.0 659 3.820299 6 240 3.0 662 2.949655 6 256 3.0 670 2.8885822 6 266 3.0 676 3.1707642 6 356 5.0 725 4.626066 6 359 3.0 728 2.949655 6 374 3.0 738 2.9234004 6 377 5.0 739 4.290662 6 416 3.0 749 2.949655 6 419 3.0 750 1.6574447 6 445 3.0 757 3.1792042 6 458 3.0 762 2.949655 6 508 4.0 782 3.8630586 6 536 3.0 792 2.949655 6 593 4.0 809 3.9737954 6 710 3.0 831 2.949655 6 818 3.0 845 2.949655 6 835 4.0 847 3.9328732 6 838 4.0 849 3.5096846 6 842 3.0 851 2.949655 6 1084 3.0 873 3.0113034 7 165 4.0 878 3.8180413 7 595 3.5 889 3.3193603 7 648 4.0 890 1.9831765 7 1610 4.0 907 4.0254817 7 1784 0.5 910 0.994102 7 1954 4.5 913 4.1282334 7 3114 4.5 927 4.1458745 7 3578 1.5 930 1.9259014 7 5816 4.5 960 4.125039 7 8368 2.5 976 2.4350963 7 8373 2.0 977 2.0334005 7 8798 4.5 983 4.477821 7 8958 3.5 991 3.544245 7 42002 4.5 1013 4.40435 7 45499 4.0 1014 3.7290354 8 11 4.0 1028 3.804239 8 110 3.0 1035 3.056051 8 454 4.0 1061 3.7677813 8 539 3.0 1066 3.0696223 8 586 3.0 1067 1.5774461 9 223 4.0 1075 3.936657 9 627 3.0 1077 2.9547868 9 1987 2.0 1085 1.9496098 9 2012 4.0 1087 3.7178297 9 3735 3.0 1094 2.9244146 9 5891 1.0 1109 0.9748049 10 4995 4.0 1141 3.8984234 10 5952 4.0 1146 3.880145 10 7151 3.0 1154 3.0667233 10 8961 2.5 1168 2.648744 10 51834 1.5 1185 1.7496729 10 59333 2.5 1191 2.4442234 10 59421 2.0 1192 1.9553789 10 63113 3.5 1196 3.4153728 10 79091 5.0 1216 4.780376 10 84374 3.5 1223 3.421913 10 88163 4.0 1226 3.9168868 10 98203 3.0 1239 2.933068 10 140110 5.0 1258 4.749053 11 377 3.0 1276 3.063622 11 466 3.0 1280 1.9338841 11 529 5.0 1285 4.82746 11 589 4.0 1286 3.9866476 11 1586 4.0 1305 3.8252048 12 2072 5.0 1339 4.9251943 13 597 3.0 1357 2.9868426 13 1590 4.0 1360 3.8396158 13 3793 4.0 1375 3.9469142 14 19 1.0 1388 2.7283285 14 47 5.0 1392 3.3924742 14 105 3.0 1394 3.0261958 14 434 4.0 1414 3.833696 14 508 3.0 1420 2.8157609 15 2012 4.0 1468 3.8306067 15 2085 1.5 1471 1.6741232 15 3949 4.0 1489 3.9425259 15 4022 3.5 1492 3.402268 15 33493 5.0 1513 4.7479286 15 71264 2.5 1529 2.5025866 15 115149 2.0 1552 3.7572365 15 122886 5.0 1556 4.6143384 15 122904 2.0 1557 2.0547605 15 152081 3.0 1564 3.0686069 15 166635 3.5 1568 3.460845 16 904 3.5 1588 4.087707 16 1197 4.0 1600 3.9888422 16 1213 3.5 1607 4.2572055 16 1252 4.5 1614 4.389011 16 3000 4.0 1636 4.1448827 16 7099 3.5 1655 2.8921077 17 44 3.5 1668 3.2983541 17 150 3.5 1673 3.5700128 17 858 5.0 1688 4.9610834 17 1209 3.5 1704 3.6166995 17 1213 4.5 1706 4.442084 17 1221 4.5 1707 3.7113688 17 1222 4.5 1708 4.4316735 17 1259 4.0 1714 3.8878596 17 2571 5.0 1730 4.595609 17 2717 3.5 1731 3.3793118 17 3702 3.5 1741 3.5332992 17 7153 4.5 1754 4.622589 17 80463 4.0 1769 3.8674312 18 555 4.0 1819 3.935853 18 780 4.0 1833 3.6659644 18 908 4.0 1838 3.9047282 18 1375 3.5 1885 3.46

In [55]:
%sql select * from movies join alldata on movies.movieId=alldata.movieId order by id

movieId title genres userId movieId rating id prediction 50 Usual Suspects, The (1995) Crime|Mystery|Thriller 1 50 5.0 4 4.8139777 673 Space Jam (1996) Adventure|Animation|Children|Comedy|Fantasy|Sci-Fi 1 673 3.0 39 3.0140967 1298 Pink Floyd: The Wall (1982) Drama|Musical 1 1298 5.0 90 4.787555 1408 Last of the Mohicans, The (1992) Action|Romance|War|Western 1 1408 3.0 94 3.1122477 2596 SLC Punk! (1998) Comedy|Drama 1 2596 5.0 168 4.868236 2616 Dick Tracy (1990) Action|Crime 1 2616 4.0 169 3.8497894 2628 Star Wars: Episode I - The Phantom Menace (1999) Action|Adventure|Sci-Fi 1 2628 4.0 171 1.4667412 2797 Big (1988) Comedy|Drama|Fantasy|Romance 1 2797 4.0 182 4.071091 3034 Robin Hood (1973) Adventure|Animation|Children|Comedy|Musical 1 3034 5.0 199 4.866223 3053 Messenger: The Story of Joan of Arc, The (1999) Drama|War 1 3053 5.0 201 2.9049928 3439 Teenage Mutant Ninja Turtles II: The Secret of the Ooze (1991) Action|Children|Fantasy 1 3439 4.0 211 1.7064148 318 Shawshank Redemption, The (1994) Crime|Drama 2 318 3.0 232 3.6837125 333 Tommy Boy (1995) Comedy 2 333 4.0 233 3.9403274 86345 Louis C.K.: Hilarious (2010) Comedy 2 86345 4.0 249 3.8730474 91658 Girl with the Dragon Tattoo, The (2011) Drama|Thriller 2 91658 2.5 252 2.4990103 1587 Conan the Barbarian (1982) Action|Adventure|Fantasy 3 1587 4.5 275 4.405271 70946 Troll 2 (1990) Fantasy|Horror 3 70946 5.0 298 4.959308 319 Shallow Grave (1994) Comedy|Drama|Thriller 4 319 5.0 321 -0.6944257 351 Corrina, Corrina (1994) Comedy|Drama|Romance 4 351 3.0 325 3.009282 902 Breakfast at Tiffany's (1961) Drama|Romance 4 902 4.0 350 0.32799235 910 Some Like It Hot (1959) Comedy|Crime 4 910 5.0 353 4.625452 919 Wizard of Oz, The (1939) Adventure|Children|Fantasy|Musical 4 919 5.0 356 4.657628 1060 Swingers (1996) Comedy|Drama 4 1060 2.0 363 2.0111842 1197 Princess Bride, The (1987) Action|Adventure|Comedy|Fantasy|Romance 4 1197 5.0 377 4.4705815 1199 Brazil (1985) Fantasy|Sci-Fi 4 1199 2.0 379 2.1057947 1211 Wings of Desire (Himmel über Berlin, Der) (1987) Drama|Fantasy|Romance 4 1211 1.0 381 1.1924 1213 Goodfellas (1990) Crime|Drama 4 1213 4.0 382 3.7148514 1732 Big Lebowski, The (1998) Comedy|Crime 4 1732 4.0 406 3.8900495 1966 Metropolitan (1990) Comedy 4 1966 2.0 420 1.9717973 2351 Nights of Cabiria (Notti di Cabiria, Le) (1957) Drama 4 2351 4.0 436 3.9435947 2571 Matrix, The (1999) Action|Sci-Fi|Thriller 4 2571 1.0 442 1.5312067 2858 American Beauty (1999) Drama|Romance 4 2858 5.0 454 2.8146186 3175 Galaxy Quest (1999) Adventure|Comedy|Sci-Fi 4 3175 1.0 467 1.1713831 3358 Defending Your Life (1991) Comedy|Drama|Fantasy|Romance 4 3358 3.0 472 3.0967898 364 Lion King, The (1994) Adventure|Animation|Children|Drama|Musical|IMAX 5 364 3.0 540 2.9546652 592 Batman (1989) Action|Crime|Thriller 5 592 3.0 554 2.9313312 13 Balto (1995) Adventure|Animation|Children 6 13 3.0 569 3.0122824 16 Casino (1995) Crime|Drama 6 16 4.0 571 4.0046167 60 Indian in the Cupboard, The (1995) Adventure|Children|Fantasy 6 60 4.0 591 3.804031 66 Lawnmower Man 2: Beyond Cyberspace (1996) Action|Sci-Fi|Thriller 6 66 3.0 595 2.9374733 76 Screamers (1995) Action|Sci-Fi|Thriller 6 76 4.0 596 3.517944 100 City Hall (1996) Drama|Thriller 6 100 3.0 605 2.966446 113 Before and After (1996) Drama|Mystery 6 113 3.0 611 2.949655 145 Bad Boys (1995) Action|Comedy|Crime|Drama|Thriller 6 145 4.0 616 3.9295845 153 Batman Forever (1995) Action|Adventure|Comedy|Crime 6 153 3.0 620 2.9510832 177 Lord of Illusions (1995) Horror 6 177 3.0 631 2.949655 185 Net, The (1995) Action|Crime|Thriller 6 185 3.0 635 3.0158434 209 White Man's Burden (1995) Drama 6 209 3.0 646 2.949655 230 Dolores Claiborne (1995) Drama|Thriller 6 230 5.0 656 4.768962 234 Exit to Eden (1994) Comedy 6 234 2.0 658 2.086646 236 French Kiss (1995) Action|Comedy|Romance 6 236 4.0 659 3.820299 240 Hideaway (1995) Thriller 6 240 3.0 662 2.949655 256 Junior (1994) Comedy|Sci-Fi 6 256 3.0 670 2.8885822 266 Legends of the Fall (1994) Drama|Romance|War|Western 6 266 3.0 

## Recommend moive to users with id: 575, 232. 
you can choose some users to recommend the moives

In [57]:
#recommend top 3 movies for each user
rec = best_model.recommendForAllUsers(3)
rec.show()

+------+--------------------+
userId| recommendations|
+------+--------------------+
 471|[[858, 3.6844938]...|
 463|[[1968, 4.37465],...|
 496|[[1214, 4.353657]...|
 148|[[40629, 4.786015...|
 540|[[1242, 5.2324915...|
 392|[[1242, 5.328126]...|
 243|[[5418, 5.380747]...|
 31|[[3153, 5.034675]...|
 516|[[1242, 4.015292]...|
 580|[[8371, 4.9335184...|
 251|[[1204, 5.335689]...|
 451|[[247, 4.3191214]...|
 85|[[66934, 4.112015...|
 137|[[1299, 4.8460994...|
 65|[[3972, 5.1979895...|
 458|[[1358, 5.599806]...|
 481|[[1729, 3.9102905...|
 53|[[4019, 4.941082]...|
 255|[[1194, 4.7605963...|
 588|[[1242, 4.350664]...|
+------+--------------------+
only showing top 20 rows

In [58]:
rec.count()

Out[35]: 595

In [59]:
rec.where(rec.userId == 575).collect()

Out[36]: [Row(userId=575, recommendations=[Row(movieId=2568, rating=3.916980743408203), Row(movieId=60756, rating=3.2804129123687744), Row(movieId=1721, rating=3.0776519775390625)])]

In [60]:
%sql select * from movies where movieId in (2568, 60756, 1721) order by movieId

movieId title genres 1721 Titanic (1997) Drama|Romance 2568 Mod Squad, The (1999) Action|Crime 60756 Step Brothers (2008) Comedy

In [61]:
rec.where(rec.userId == 232).collect()

Out[37]: [Row(userId=232, recommendations=[Row(movieId=1214, rating=4.8962788581848145), Row(movieId=2959, rating=4.895084381103516), Row(movieId=2924, rating=4.895073890686035)])]

In [62]:
%sql select * from movies where movieId in (1214, 2959, 2924) order by movieId

movieId title genres 1214 Alien (1979) Horror|Sci-Fi 2924 Drunken Master (Jui kuen) (1978) Action|Comedy 2959 Fight Club (1999) Action|Crime|Drama|Thriller

## Find the similar movies for movies with id: 463, 471
You can find the similar moives based on the ALS results

In [64]:
from pyspark.ml.linalg import Vectors
from pyspark.sql.functions import col
from pyspark.ml.feature import BucketedRandomProjectionLSH
from pyspark.ml.linalg import Vectors, VectorUDT

In [65]:
%sql select * from movies where movieId = 471

movieId title genres 471 Hudsucker Proxy, The (1994) Comedy

In [66]:
a = best_model.itemFactors
display(a.cache())

id features 10 List(-0.17399456, -0.11281259, 0.8569175, -0.54631084, -0.55400944, 1.4655101, -0.64992434, 0.11219682, -0.58871645, -0.805032) 20 List(0.48783374, 0.26239246, 1.025266, -0.8355324, -0.64349926, 1.1789007, -0.38052866, 0.12471497, -0.36214808, 0.46206874) 40 List(0.011918607, -0.6823519, 1.2163125, 0.6471562, -1.8250132, -0.11542535, -1.0401864, -0.16205989, 0.13153061, -0.6745505) 50 List(0.41380128, 0.52764153, 0.3219214, -0.59345657, -0.8761349, 1.5146528, -0.88337696, 0.8208594, -1.0010954, -0.6537329) 60 List(-0.13563386, 0.9672694, 0.74212706, -0.73763955, -0.8961581, 1.0795335, -0.14413597, 0.6976324, -0.31207517, -0.4542721) 70 List(0.1868956, 0.11975827, 0.24218467, 0.28644493, -1.4317477, 1.8521532, -1.1187931, 0.25387195, -0.060563233, -0.38064763) 100 List(0.3087319, 0.5037343, 0.49833703, -0.14110358, -0.7671431, 0.92059064, 0.051934868, 0.93045247, 0.36248496, -0.062180847) 110 List(0.60796386, 0.70030755, 1.4684035, 0.12191183, -0.30424312, 1.4313833, -0.60322106, 0.3105281, -0.90391064, -0.6063106) 140 List(-0.4878867, 0.6201137, -0.76593447, -0.5902814, 0.33427775, 0.33656263, -1.7366278, 0.20049278, 0.20113528, -1.1813973) 150 List(0.5027478, -0.07749295, 0.5137244, -0.25358272, -1.2210498, 1.7142764, -0.7389047, 0.6288325, 0.05641706, -1.2028278) 160 List(0.1594956, 0.6075989, 1.453608, 0.026553962, -0.17478178, 1.2532855, -0.061089218, -0.23652498, 0.107782856, -0.3038706) 170 List(0.70871896, -0.051804785, 1.1358262, -0.34158227, 0.10897925, 1.4248914, -0.37947503, 0.8311732, -0.3585489, 0.38110852) 180 List(0.09088633, -0.25320795, 0.9437433, -0.31845766, -0.32807475, 0.50784445, -0.8108974, 1.1711463, 0.30811244, -0.24388838) 230 List(0.38761985, 0.67083436, 1.0550882, -0.7101561, -1.0790049, 1.4519314, -0.0427187, 1.0064793, 0.030096151, -0.4172677) 240 List(0.24502338, 0.41271973, 0.66136396, -0.43852085, -0.6653826, 0.9026198, -0.02995972, 0.6111468, -1.9075438E-4, -0.24571137) 260 List(0.1254836, 0.29762396, 0.86947054, -1.1097564, -0.45875958, 1.7191558, -1.6817749, -0.2835234, -0.063310154, -0.5896467) 280 List(0.3416126, 0.4509361, 1.1491752, -0.71941495, -0.8849182, 1.6448874, -0.7902279, 0.6873439, -0.3983718, -0.18345916) 290 List(-0.07216962, 0.17108089, -0.6073082, -0.20093179, -1.3004926, 1.9412541, -0.42492473, 0.6944765, -0.5259925, -0.5998473) 300 List(0.2775346, -0.067917764, 0.73555756, -0.5942547, -0.9525655, 0.92286026, -1.01373, 0.44047496, -0.24597026, -0.5669428) 350 List(0.7369044, 0.1230462, 1.2967002, -0.9570597, -0.05632071, 0.8526861, 0.38244694, -0.29764065, 0.27871823, -0.23043059) 360 List(0.43809572, -0.026541369, 0.19848405, -0.2010997, 0.25122824, 1.0803033, -0.3213993, 1.0794486, -0.107024334, 0.34563097) 370 List(1.2954968, 0.16875558, 1.1675099, -0.75508076, 0.034368474, 0.8143432, 0.5124267, 0.079048894, -0.4012926, -0.07373844) 380 List(-0.091042645, -0.19490376, 0.82518405, 0.064407, -0.58775383, 1.7711097, -0.65667117, 0.1885111, -0.73863804, -0.57536083) 410 List(0.38622817, 0.2951013, 1.0062399, -0.20616354, -0.50702906, 1.2998346, -0.6004896, 0.41795084, 0.35863882, -0.054803208) 420 List(0.07991673, 0.25100273, 0.918372, -0.5339287, -0.8094243, 0.89688754, -0.651678, -0.08958508, -0.45550746, 0.25906762) 440 List(0.6871257, 0.32536846, 0.97726655, -0.1186698, -0.46104732, 1.683458, -0.4759724, 0.33681178, -0.71562314, -0.3118114) 450 List(-0.11407638, 0.17477623, -0.27452627, -0.25686952, -0.9774206, 1.6021571, -0.38581055, 0.525637, -0.30871293, -0.48764554) 480 List(0.33141842, 0.2839652, 1.3687856, -0.06398181, 0.071738325, 1.7069398, -0.5215541, 0.051962778, -1.2562873, -0.7975262) 500 List(-0.0010691802, 0.16518764, 1.0378054, -1.3446871, -0.51225173, 1.1746926, -0.58351046, 0.17509826, -0.41922712, -0.2044558) 520 List(-0.27261037, -0.8246396, 1.1307902, 0.14862625, -0.57413065, 0.5317573, -0.4487846, 0.6740432, -0.9720638, -0.3555339) 540 List(0.60264933, 0.80859613, 0.5677964, 0.14196981, -0.018823313, 0.6766598, 0.2305603, 0.8011426, -0

In [67]:
a.where(a.id == 471).collect()

Out[40]: [Row(id=471, features=[0.41469210386276245, -0.00914978887885809, 1.2196894884109497, -1.4785195589065552, 0.542229175567627, 1.3933889865875244, 0.11748117208480835, 0.44705721735954285, 0.40512022376060486, 0.3802637755870819])]

In [68]:
a.where(a.id == 463).collect()

Out[41]: []

In [69]:
a.registerTempTable("itemMatrix")

In [70]:
%sql
select features from itemMatrix where id = 471

features List(0.4146921, -0.009149789, 1.2196895, -1.4785196, 0.5422292, 1.393389, 0.11748117, 0.44705722, 0.40512022, 0.38026378)

In [71]:
%sql
select features from itemMatrix where id = 463

features

In [72]:
brp = BucketedRandomProjectionLSH(inputCol="features", outputCol="hashes",seed=12345, bucketLength=1.0)
a.printSchema()
#change features columns into dense vector
to_vector = udf(lambda a: Vectors.dense(a), VectorUDT())
data = a.select("id", to_vector("features").alias("features"))
data.printSchema()

root
-- id: integer (nullable = false)
-- features: array (nullable = true)
 |-- element: float (containsNull = false)

root
-- id: integer (nullable = false)
-- features: vector (nullable = true)

In [73]:
model = brp.fit(data)
model.transform(data)

Out[59]: DataFrame[id: int, features: vector, hashes: array<vector>]

In [74]:
model.approxNearestNeighbors(data, Vectors.dense([0.4146921,-0.009149789,1.2196895,-1.4785196,0.5422292,1.393389,0.11748117,0.44705722,0.40512022,0.38026378]), 6).collect()

Out[62]: [Row(id=471, features=DenseVector([0.4147, -0.0091, 1.2197, -1.4785, 0.5422, 1.3934, 0.1175, 0.4471, 0.4051, 0.3803]), hashes=[DenseVector([0.0])], distCol=5.157211161170817e-08),
 Row(id=100383, features=DenseVector([0.4584, 0.1605, 0.6619, -1.1362, 0.9767, 1.6138, -0.2073, 0.716, -0.249, 0.2862]), hashes=[DenseVector([0.0])], distCol=1.1449081505244467),
 Row(id=4226, features=DenseVector([0.0375, 0.1906, 1.0406, -1.1876, 0.1527, 1.9812, -0.3894, 0.9659, 0.082, 0.3408]), hashes=[DenseVector([0.0])], distCol=1.1950648132422377),
 Row(id=1500, features=DenseVector([1.0099, 0.0422, 0.6341, -1.4007, -0.1794, 1.5966, 0.3951, 0.4336, 0.0372, 0.4171]), hashes=[DenseVector([0.0])], distCol=1.2173377910514922),
 Row(id=5957, features=DenseVector([0.3469, 0.1886, 0.6857, -1.1686, -0.0978, 1.6653, -0.3737, 0.3267, -0.1077, 0.1119]), hashes=[DenseVector([0.0])], distCol=1.2244565522772075),
 Row(id=246, features=DenseVector([0.6531, 0.2292, 1.4375, -1.4201, -0.3641, 1.5474, 0.0866, 0.5251, -0.3328, 0.4547]), hashes=[DenseVector([0.0])], distCol=1.2517385838135278)]

In [75]:
%sql
select * from movies
where movieId in (100383,4226,1500,5957,246)

movieId title genres 246 Hoop Dreams (1994) Documentary 1500 Grosse Pointe Blank (1997) Comedy|Crime|Romance 4226 Memento (2000) Mystery|Thriller 5957 Two Weeks Notice (2002) Comedy|Romance 100383 Side Effects (2013) Crime|Drama|Mystery|Thriller

## Write the report 
motivation
1. step1
2. step2
3. step3
4. step4  
output and conclusion

## Movie Recommendation with Spark ALS
1. Use Alternating Least Square with Spark to predict ratings for movies in MovieLens small dataset
2. Perform Data ETL and Data Exploration to learn basic information and prepare data for analysis
3. Analyze datasets in terms of movies, users, ratings and genres via Spark SQL and OLAP
4. Train an ALS model on a dataset of 10000 data points, tune hyperparameters using cross validation to get best model, use best model to predict ratings and achieve 0.56 RMSE, recommend top 3 movies to users, and find similar movies for movies using ALS’s item matrix and k-nearest neighbors